tachyon/zk/plonk/vanishing/prover_vanishing_argument.h

```diff
@@ -0,0 +1,322 @@
+    CHECK_EQ(evals.NumElements(), extended_domain->size());
+
+    // Calculate ζ = g^((2ˢ * T) / 3)^2
+    // NOTE(TomTaehoonKim): Halo2 does not calculate ζ = g^((2ˢ * T) / 3), but
+    // the square of it.
+    CHECK_EQ(F::Config::kTrace % math::BigInt<F::kLimbNums>(3),
+             math::BigInt<F::kLimbNums>(0));
+    F zeta = F::FromMontgomery(F::Config::kSubgroupGenerator)
+                 .Pow(F(2).Pow(F::Config::kTwoAdicity).ToBigInt() *
+                      F::Config::kTrace / math::BigInt<F::kLimbNums>(3))
+                 .Square();
+
+    // Compute the evaluations of t(X) = Xⁿ - 1 in the coset evaluation domain.
+    // We don't have to compute all of them, because it will repeat.
+    std::vector<F> t_evaluations;
+    t_evaluations.reserve(size_t{1} << (extended_domain->log_size_of_group() -
+                                        domain->log_size_of_group()));
+    F orig = zeta.Pow(domain->size());
+    F step = extended_domain->group_gen().Pow(domain->size());
+    F cur = orig;
+    do {
+      t_evaluations.push_back(cur);
+      cur *= step;
+    } while (cur != orig);
+    CHECK_EQ(t_evaluations.size(),
+             size_t{1} << (extended_domain->log_size_of_group() -
+                           domain->log_size_of_group()));
+
+    // Subtract 1 from each to give us t_evaluations[i] = t(zeta *
+    // extended_omegaⁱ)
+    base::Parallelize(t_evaluations, [](absl::Span<F> chunk) {
+      for (F& coeff : chunk) {
+        coeff -= F::One();
+      }
+    });
+
+    F::BatchInverseInPlace(t_evaluations);
+
+    // Multiply the inverse to obtain the quotient polynomial in the coset
+    // evaluation domain.
+    // TODO(TomTaehoonKim): Avoid copy here by adding element wise
+    // multiplication.
+    std::vector<F> evaluations = evals.evaluations();
+    base::Parallelize(evaluations,
+                      [t_evaluations](absl::Span<F> chunk, size_t chunk_idx,
+                                      size_t chunk_size) {
+                        size_t index = chunk_idx * chunk_size;
+                        for (F& h : chunk) {
+                          h *= t_evaluations[index % t_evaluations.size()];
+                          ++index;
+                        }
+                      });
+
+    return ExtendedEvals(std::move(evaluations));
+  }
+
+  // This takes us from the extended evaluation domain and gets us the quotient
+  // polynomial coefficients.
+  //
+  // This function will panic if the provided vector is not the correct length.
+  constexpr ExtendedPoly ExtendedToCoeff(
+      const ExtendedEvals& evals, const ExtendedDomain* extended_domain) const {
+    CHECK_EQ(evals.NumElements(), extended_domain->size());
+
+    ExtendedPoly poly = extended_domain->IFFT(evals);
+
+    // Distribute powers to move from coset; opposite from the
+    // transformation we performed earlier.
+    DistributePowersZeta(poly, false);
+
+    return poly;
+  }
```

Comment:
It may static method. (but not marked)
And we also need to discuss the placement of this function. @chokobole 
That's why, it will be used in other scope (ex, `EvaluteH()`)